File: YourMainActivity.java
package com.your.package;

import android.content.SharedPreferences;
import android.os.Build;
import android.os.Bundle;
import android.app.Activity;
import android.widget.TextView;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.FileReader;
import java.net.HttpURLConnection;
import java.net.URL;

public class YourMainActivity extends Activity {
  private SharedPreferences pref;
  
  @Override
protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    pref = getSharedPreferences("sp", Activity.MODE_PRIVATE);

    String cpuCode = getCPUInfo();
    getChipsetData(cpuCode);
}

private String getCPUInfo() {
    try {
        BufferedReader br = new BufferedReader(new FileReader("/proc/cpuinfo"));
        String line;
        StringBuilder cpuInfo = new StringBuilder();
        while ((line = br.readLine()) != null) {
            if (line.toLowerCase().contains("hardware") ||
                line.toLowerCase().contains("model name")) {
                cpuInfo.append(line.split(":")[1].trim());
                break;
            }
        }
        br.close();
        String raw = cpuInfo.length() > 0 ? cpuInfo.toString() : Build.HARDWARE;
        return raw.toUpperCase().replaceAll("[^A-Z0-9]", "");
    } catch (Exception e) {
        return Build.HARDWARE.toUpperCase().replaceAll("[^A-Z0-9]", "");
    }
}

private void getChipsetData(String cpuCode) {
    String cachedCpu = pref.getString(cpuCode+"_cpu", null);
    String cachedGpu = pref.getString(cpuCode+"_gpu", null);

    if(cachedCpu != null && cachedGpu != null){
        tvCPU.setText(cachedCpu);
        tvGpu.setText(cachedGpu);
        return;
    }

    String url = "";
    if(cpuCode.startsWith("MT")){
        url = "https://tolepcoy.pages.dev/tolepcoy_mediatek?code=" + cpuCode;
    } else if(cpuCode.startsWith("MSM") || cpuCode.startsWith("SM")){
        url = "https://tolepcoy.pages.dev/tolepcoy_qualcomm?code=" + cpuCode;
    } else {
        tvCPU.setText("Chipset not found!");
        tvGpu.setText("Unknown");
        return;
    }

    String finalUrl = url;
    new Thread(() -> {
        try {
            URL urlObj = new URL(finalUrl);
            HttpURLConnection conn = (HttpURLConnection) urlObj.openConnection();
            conn.setRequestMethod("GET");

            BufferedReader br = new BufferedReader(new InputStreamReader(conn.getInputStream()));
            StringBuilder result = new StringBuilder();
            String line;
            while((line = br.readLine()) != null){
                result.append(line);
            }
            br.close();

            JSONObject json = new JSONObject(result.toString());

            String cpu = json.optString("chipset", "Nothing");
            String gpu = json.optString("gpu", "Nothing");

            // Save to SharedPreferences
            pref.edit()
                .putString(cpuCode+"_cpu", cpu)
                .putString(cpuCode+"_gpu", gpu)
                .apply();

            runOnUiThread(() -> {
                tvCPU.setText(cpu);
                tvGpu.setText(gpu);
            });

        } catch(Exception e){
            runOnUiThread(() -> {
                tvCPU.setText("Chipset not found!");
                tvGpu.setText("Unknown");
            });
        }
    }).start();
}
}

File: YourMainActivity.kt
package com.your.package

import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.app.Activity
import android.widget.TextView
import org.json.JSONObject
import java.io.BufferedReader
import java.io.FileReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

class YourMainActivity : Activity() {

    private lateinit var pref: SharedPreferences
    private lateinit var tvCPU: TextView
    private lateinit var tvGpu: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        pref = getSharedPreferences("sp", Activity.MODE_PRIVATE)
        tvCPU = findViewById(R.id.tvCPU)
        tvGpu = findViewById(R.id.tvGpu)

        val cpuCode = getCPUInfo()
        getChipsetData(cpuCode)
    }

    private fun getCPUInfo(): String {
        return try {
            val br = BufferedReader(FileReader("/proc/cpuinfo"))
            val cpuInfo = StringBuilder()
            var line: String?
            while (br.readLine().also { line = it } != null) {
                line?.let {
                    val lower = it.lowercase()
                    if (lower.contains("hardware") || lower.contains("model name")) {
                        cpuInfo.append(it.split(":")[1].trim())
                        break
                    }
                }
            }
            br.close()
            val raw = if (cpuInfo.isNotEmpty()) cpuInfo.toString() else Build.HARDWARE
            raw.uppercase().replace(Regex("[^A-Z0-9]"), "")
        } catch (e: Exception) {
            Build.HARDWARE.uppercase().replace(Regex("[^A-Z0-9]"), "")
        }
    }

    private fun getChipsetData(cpuCode: String) {
        val cachedCpu = pref.getString("${cpuCode}_cpu", null)
        val cachedGpu = pref.getString("${cpuCode}_gpu", null)

        if (cachedCpu != null && cachedGpu != null) {
            tvCPU.text = cachedCpu
            tvGpu.text = cachedGpu
            return
        }

        val url = when {
            cpuCode.startsWith("MT") -> "https://tolepcoy.pages.dev/tolepcoy_mediatek?code=$cpuCode"
            cpuCode.startsWith("MSM") || cpuCode.startsWith("SM") -> "https://tolepcoy.pages.dev/tolepcoy_qualcomm?code=$cpuCode"
            else -> {
                tvCPU.text = "Chipset not found!"
                tvGpu.text = "Unknown"
                return
            }
        }

        Thread {
            try {
                val urlObj = URL(url)
                val conn = urlObj.openConnection() as HttpURLConnection
                conn.requestMethod = "GET"

                val br = BufferedReader(InputStreamReader(conn.inputStream))
                val result = StringBuilder()
                var line: String?
                while (br.readLine().also { line = it } != null) {
                    result.append(line)
                }
                br.close()

                val json = JSONObject(result.toString())
                val cpu = json.optString("chipset", "Nothing")
                val gpu = json.optString("gpu", "Nothing")

                pref.edit()
                    .putString("${cpuCode}_cpu", cpu)
                    .putString("${cpuCode}_gpu", gpu)
                    .apply()

                runOnUiThread {
                    tvCPU.text = cpu
                    tvGpu.text = gpu
                }
            } catch (e: Exception) {
                runOnUiThread {
                    tvCPU.text = "Chipset not found!"
                    tvGpu.text = "Unknown"
                }
            }
        }.start()
    }
}